File: TSP/ParserTSPFile.py
class ParserTSPFile:
    FULL_MATRIX = "FULL_MATRIX"
    LOWER_DIAG_ROW = "LOWER_DIAG_ROW"

    def __init__(self, matrixFileName, solutionFileName):
        self.matrixFileName = matrixFileName
        self.solutionFileName = solutionFileName

    def getMatrix(self):
        file = open(self.matrixFileName)
        line = file.readline().rstrip()
        if (line == ParserTSPFile.FULL_MATRIX):
            matrix = self._getFullMatrix(file)
        else:
            matrix = self._getLowerDiagRowMatrix(file)

        file.close()

        return matrix

    def _getFullMatrix(self,file):
        matrix = []
        for line in file:
            lineSplit = line.split()
            matrix.append([int(x) for x in lineSplit])

        return matrix

    def _getLowerDiagRowMatrix(self,file):
        matrix = [[]]
        actualRow = 0
        actualCol = 0
        rowLengthToRead = 1
        for line in file:
            distance = int(line)
            matrix[actualRow].append(distance)
            if (actualRow != actualCol):
                matrix[actualCol].append(distance)
            actualCol += 1
            if (actualCol >= rowLengthToRead):
                actualCol = 0
                actualRow += 1
                rowLengthToRead += 1
                matrix.append([])

        matrix.pop() # Para borrar la ultima lista vacia agregada.
        return matrix

    def getSolutionList(self):
        file = open(self.solutionFileName)
        list = []
        for line in file:
            vertex = int(line) - 1
            list.append(vertex)
        file.close()
        return list

File: Viajante/ParserTSPFile.cpp
//
// Created by kevin on 11/12/16.
//

#include <sstream>
#include "ParserTSPFile.h"

std::string ParserTSPFile::FULL_MATRIX = "FULL_MATRIX";

std::string ParserTSPFile::LOWER_DIAG_ROW = "LOWER_DIAG_ROW";

ParserTSPFile::ParserTSPFile(string matrixFileName, string solutionFileName) {
    this->matrixFile = ifstream(matrixFileName.c_str());
    if (!this->matrixFile.is_open()) {
        throw std::string("Error al abrir el archivo de la matriz: ") + matrixFileName;
    }
    this->solutionFile = ifstream(solutionFileName.c_str());
    if (!this->solutionFile.is_open()) {
        throw std::string("Error al abrir el archivo de la solucion: ") + solutionFileName;
    }
}

ParserTSPFile::~ParserTSPFile() {
    this->matrixFile.close();
    this->solutionFile.close();
}

vector<vector<CostInt>> *ParserTSPFile::getMatrix() {
    string line;
    getline(this->matrixFile, line);
    if (line == ParserTSPFile::FULL_MATRIX){
        return this->getFullMatrix();
    } else {
        return this->getLowerDiagRowMatrix();
    }
}

vector<CostInt> ParserTSPFile::splitLine(string line, char delim) {
    vector<CostInt > items;
    stringstream ss;
    ss.str(line);
    string item;
    while (getline(ss, item, delim)) {
        if (!item.empty()) {
            items.push_back(stoi(item));
        }
    }
    return items;
}

vector<vector<CostInt>> *ParserTSPFile::getFullMatrix() {
    vector<vector<CostInt >>* matrix = new vector<vector<CostInt >>();
    string line;
    while (getline(this->matrixFile, line)) {
        if (line.empty()) continue;
        vector<CostInt > row = this->splitLine(line, ' ');
        matrix->push_back(row);
    }

    return matrix;
}

vector<vector<CostInt >> *ParserTSPFile::getLowerDiagRowMatrix() {
    vector<vector<CostInt>>* matrix = new vector<vector<CostInt>>();
    matrix->push_back(vector<CostInt>());
    unsigned int actualRow = 0;
    unsigned int actualCol = 0;
    unsigned int rowLengthtoRead = 1;
    string line;
    while (getline(this->matrixFile, line)) {
        if (line.empty()) continue;
        unsigned int distance = stoi(line);
        matrix->at(actualRow).push_back(distance);
        if (actualRow != actualCol) {
            matrix->at(actualCol).push_back(distance);
        }
        actualCol += 1;
        if (actualCol >= rowLengthtoRead) {
            actualCol = 0;
            actualRow += 1;
            rowLengthtoRead += 1;
            matrix->push_back(vector<CostInt>());
        }
    }

    matrix->pop_back(); // Borrar la ultima lista vacia agregada.

    return matrix;

}

vector<VertexInt>* ParserTSPFile::getSolutionList() {
    vector<VertexInt>* lista = new vector<VertexInt>();
    string line;
    while (getline(this->solutionFile, line)) {
        if (line.empty()) continue;
        VertexInt vertex = stoi(line) - 1;
        lista->push_back(vertex);
    }

    return lista;
}
